spec/support/cmap.rb

```diff
@@ -0,0 +1,346 @@
+                        'type' => 'ConnectionCreated',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionReady
+                      {
+                        'type' => 'ConnectionReady',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionClosed
+                      {
+                        'type' => 'ConnectionClosed',
+                        'connectionId' => event.connection_id,
+                        'reason' => event.reason,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionCheckoutStarted
+                      {
+                        'type' => 'ConnectionCheckOutStarted',
+                    }
+                    when Mongo::Monitoring::Event::ConnectionCheckoutFailed
+                      {
+                        'type' => 'ConnectionCheckOutFailed',
+                        'reason' => event.reason,
+                      }
+                   when Mongo::Monitoring::Event::ConnectionCheckedOut
+                      {
+                        'type' => 'ConnectionCheckedOut',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionCheckedIn
+                      {
+                        'type' => 'ConnectionCheckedIn',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::PoolCleared
+                      {
+                        'type' => 'ConnectionPoolCleared',
+                        'address' => event.address,
+                      }
+                    else
+                      next
+                    end
+
+            events << event unless @ignore.include?(event['type'])
+            events
+          end
+        end
+      end
+
+      private
+
+      # Converts the options used by the Ruby driver to the spec test format.
+      def normalize_options(options)
+        (options || {}).reduce({}) do |opts, kv|
+         case kv.first
+          when :max_idle_time
+            opts['maxIdleTimeMS'] = (kv.last * 1000.0).to_i
+          when :max_pool_size
+            opts['maxPoolSize'] = kv.last
+          when :min_pool_size
+            opts['minPoolSize'] = kv.last
+          when :wait_queue_size
+            opts['waitQueueSize'] = kv.last
+          when :wait_queue_timeout
+            opts['waitQueueTimeoutMS'] = (kv.last * 1000.0).to_i
+          end
+
+          opts
+        end
+      end
+
+      # Converts the options given by the spec to the Ruby driver format.
+      def process_options(options)
+        (options || {}).reduce({}) do |opts, kv|
+          case kv.first
+          when 'maxIdleTimeMS'
+            opts[:max_idle_time] = kv.last / 1000.0
```

Comment:
Are these URI options, and if so, isn't there any code in the driver which would convert the URI options to driver options that can be reused and potentially be made available to users as well?

Reply:
I think we should file a ticket for this rather than doing it here; this spec implementation has already grown a lot in scope, and we need to start winding down on it rather than adding more to it